### PR TITLE
top-level: simplify stdenv calling of top-level: fewer inferred and re-inferring arguments

### PR DESCRIPTION
**Pull request by Ericson2314**
### New summery

Ok, this is last general purpose cleanup from me before going into the nitty-gritty details of cross compiling. The basic issue is despite my changes so far *almost* but not quite break the top-level stdenv cycle. This breaks the cycle, *and* cuts down on the partial application magic that can make it hard to figure out where an argument comes from.

This PR has been tested with my usual gist
```sh
nix-instantiate --eval -E 'import (builtins.fetchTarball https://gist.github.com/Ericson2314/fc745290db68b20235bdc457961b3658/archive/master.tar.gz) {}' -A testsPass
```
now extended to also check ghc's linux and darwin hashes in addition to cross + custom stdenv stuff. It should also pass travis--I ran the nox script locally.

In the process of cleaning up the default args used for testing, I added a builtins trace as they should never be triggered in production.

### Original Post

This grew out of the, ahem, tamer of my ideas in #10874. The basic idea is that `top-level/default.nix` is now only responsible for building a single bootstrapping phase, and since stdenvs control the bootstrapping, they are solely responsible for calling `top-level/default.nix` (well, with the exception of the final phase, since that isn't composed in a stdenv-specific manner). Additionally config's `replaceStdenv` and the cross stdenv are now composed more like other stdenvs than the ad-hoc way they were before.

While this doesn't really solve any problem on its own, it should lay a better foundation for feature work. See the comments in #14965 for how the phasing for the cross stdenv helps for example. Also, some of the more ambitious proposals of #10874 should similarly be assisted.

I tested (a few arbitrary) package paths stayed the same on Nixos. I verified the same for cross-compiling before I rebased onto the newly-bumped nixos-unstable, but post-rebase, cross-compiling seems to be broken for reasons unrelated to this.

The biggest problem with this is now the pkg hierarchy makes less sense: `top-level` isn't really the top-level and theirs a few more wrapper files. I'd gladly move things around (`phase-top-level`?) but not sure what the best way forward is.

I view this the spiritual successor of #14000, and as such, I tried to mimic @nbp's method of breaking this up into little commits to make it easier to review.

CC @vcunat @Mathnerd314 @nbp 

[PS: @nbp while you weren't involved in the #10874 discussion, this (hopefully just syntactically) conflicts a fair amount with @10874 hence the CC.]

### CODE REVIEW

**Comment by domenkozar** [comment]
If we keep improving top-level functions (which I fully support), we should also document the flow so that we can reason about the improvements.

It's really hard to evaluate and follow the logic without spending considerable amount of time on it.

**Comment by Ericson2314** [comment]
Yeah I fully agree. I tried to add some documentation as I went along, but it would be good to move things into saner locations and add more.

File: maintainers/scripts/dep-licenses.sh [hunk 1]
```diff
@@ -38,10 +38,10 @@ if test $(wc -l "$tmp/filter.sed" | sed 's/ .*//') == 0; then
...
 No derivation mentionned in the stack trace.  Either your derivation does
 not use stdenv.mkDerivation or you forgot to use the stdenv adapter named
-traceDrvLicenses.
+traceDrvLicenses. In pkgs/top-level/stdenv.nix:
 
--  defaultStdenv = allStdenvs.stdenv;
-+  defaultStdenv = traceDrvLicenses allStdenvs.stdenv;
+-  defaultStdenv = vanillaStdenv;
++  defaultStdenv = traceDrvLicenses avanillaStdenv;
 "
```
Comment (Mathnerd314) [inline maintainers/scripts/dep-licenses.sh]: Could this script be changed to use ``nix-instantiate -E``? Then this message would not be necessary.
Also, @nbp, is this script still useful?

**Comment by Mathnerd314** [comment]
As far as the locations, I think ``pkgs/default.nix`` should be merged back into ``pkgs/top-level/default.nix`` (pure default arguments are usually helpful), and then ``pkgs/impure.nix`` can be moved to ``top-level`` as well, or else merged into the ``default.nix`` in the root directory (I can't say top-level, because that would be confusing).

File: pkgs/top-level/default.nix [hunk 2]
```diff
@@ -19,83 +25,25 @@
                && system != "x86_64-solaris"
                && system != "x86_64-kfreebsd-gnu")
```
Comment (Mathnerd314) [inline pkgs/top-level/default.nix]: ``noSysDirs`` should be in stdenv (true in stdenvNative and false elsewhere), rather than being a top-level argument.

File: pkgs/stdenv/linux/default.nix [hunk 3]
```diff
@@ -4,10 +4,12 @@
 # compiler and linker that do not search in default locations,
 # ensuring purity of components produced by it.
 
-# The function defaults are for easy testing.
-{ system ? builtins.currentSystem
-, allPackages ? import ../../..
-, platform ? null, config ? {}, lib ? (import ../../../lib)
+# Use stdenv parameter of `pkgs/top-level.nix` for easy testing.
+{ system
```
Comment (Mathnerd314) [inline pkgs/stdenv/linux/default.nix]: ``pkgs/top-level.nix`` does not exist.

Comment (Ericson2314) [inline pkgs/stdenv/linux/default.nix]: Haha, fixed

**Comment by Ericson2314** [comment]
@Mathnerd314 `pkgs/default.nix` composes the whole thing, while `pkgs/top-level/default.nix` composes just one stage --- its not just a matter of default arguments.

How about:
 - `pkgs/top-level/default.nix`  -> `pkgs/top-level/stage.nix`
 - `pkgs/default.nix`  -> `pkgs/stage.nix`

I do like the idea of just sticking the impure stuff in `/defaut.nix` :)

File: pkgs/top-level/default.nix (hunk 2 above)
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: I like that idea, but I see it less as something that should have been done already than something that's good to do in addition---I haven't messed with any parameters that actual packages receive yet.

That said, I'm fine just making it an additional commit that's part of this rather than a separate PR.

Comment (Ericson2314) [inline pkgs/top-level/default.nix]: I have a similar idea for `crossSystem`. Now that `crossStdenv` is properly staged, we can give each stdenv a `host` and `target` system+platform, and set them accordingly in the cross stages.

`crossStdenv` would remain as a entry-point argument, but `all-packages.nix` would not have that argument as just look at `stdenv.host` and `stdenv.target`.

Comment (Ericson2314) [inline pkgs/top-level/default.nix]: But that, if not the `noSysDirs` change, will probably be more contentious and deserving of its own PR.

File: maintainers/scripts/dep-licenses.sh (hunk 1 above)
Comment (Ericson2314) [inline maintainers/scripts/dep-licenses.sh]: Technically, I didn't even fix the error message to be a real diff. I was mainly interested in making sure my `allStdenvs` grep came clean before I removed it :).

**Comment by Mathnerd314** [comment]
> pkgs/default.nix composes the whole thing, while pkgs/top-level/default.nix composes just one stage --- it's not just a matter of default arguments.

I see, but that's still not a barrier to merging them. I did so [here](https://github.com/Ericson2314/nixpkgs/compare/cross...Mathnerd314:cross) ([full diff](https://github.com/NixOS/nixpkgs/compare/master...Mathnerd314:cross)), along with my other suggested changes.

**Comment by Ericson2314** [comment]
I like it! Part of me is still partial to a `default.nix` `phase.nix` separation, but it hardly matters. Note that @nbp purposely kept the throw-away-everything semantics for `pkgsWithOverrides` rather than using `__unfix__` for backwards compat, but I'm not at all sure that's preserving a feature rather than a bug. Thanks for taking a good look at this!

**Comment by Ericson2314** [comment]
Anybody else else want to review this (and @Mathnerd314's additions) soon? In some cases I imagine it might make sense to interleave/squash our commits---let me know where and I'll be happy to do it.

File: default.nix [hunk 4]
```diff
@@ -6,4 +6,4 @@ if ! builtins ? nixVersion || builtins.compareVersions requiredVersion builtins.
 
 else
 
-  import ./pkgs/top-level
+  import ./pkgs/impure.nix
```
Comment (zimbatm) [inline default.nix]: I'm not sure about the location and name about this-one but don't have a better recommendation at the moment.

File: pkgs/stdenv/darwin/default.nix [hunk 5]
```diff
@@ -15,15 +16,14 @@
     cpio    = fetch { file = "cpio";  sha256 = "0ms5i9m1vdksj575sf1djwgm7zhnvfrrb44dxnfh9avr793rc2w4"; };
     tarball = fetch { file = "bootstrap-tools.cpio.bz2"; sha256 = "1lz1b0grl4642h6n635xvi6imf0yyy1zyzdr9ing5aphzz0z5iic"; executable = false; };
   }
+, ...
```
Comment (zimbatm) [inline pkgs/stdenv/darwin/default.nix]: typo ?

**Comment by zimbatm** [comment]
Looking like a good cleanup

File: pkgs/stdenv/darwin/default.nix (hunk 5 above)
Comment (Ericson2314) [inline pkgs/stdenv/darwin/default.nix]: I don't think so. the `...`s allows us to give the all stdenvs the same arguments not worrying about which subset they actually use. See the simplifications to `stdenv/default.nix` in the same commit that introduced this.

Comment (zimbatm) [inline pkgs/stdenv/darwin/default.nix]: I see, makes sense

**Comment by Ericson2314** [comment]
Hello reviewers! :)

>  - `pkgs/top-level/default.nix`  -> `pkgs/top-level/stage.nix`
 - `pkgs/default.nix`  -> `pkgs/top-level/default.nix`

I rebased on top of latest channels/nixpkgs-unstable, did that rename I proposed, and added @Mathnerd314's commits to make it easier to review everything.

**Comment by zimbatm** [comment]
I always wondered, why is that folder called top-level when it's not the top level? Having all the packages listen in `pkgs/default.nix` would mean less typing for me. And all the `<lang>-packages.nix` could go in `pkgs/development/<lang>-modules/default.nix`. Especially the python one who just references tons of patches from that folder anyways.

**Comment by Ericson2314** [comment]
@zimbatm I agree. I tried to conservatively keep things in the same place for now. But nixpkgs is overdue for a big sorting.

File: pkgs/top-level/default.nix [hunk 6]
```diff
@@ -50,9 +48,16 @@ let
   platform = if platform_ != null then platform_
     else config.platform or platformAuto;
 
-  topLevelArguments = {
-    inherit system bootStdenv noSysDirs config crossSystem platform lib;
-  };
+  # A few packages make make a new package set to draw their dependencies from.
```
Comment (bjornfor) [inline pkgs/top-level/default.nix]: "make make" => "make"

**Comment by zimbatm** [comment]
@Ericson2314 agreed not in this PR. Speaking of which, is it ready to be merged?

**Comment by Ericson2314** [comment]
@zimbatm Well, the first *n* commits are ready, the question is what *n* is :). I am quite confident that everything through my penultimate commit is a good idea. But I'm a bit unsure whether the last commit makes things more or less convoluted. And I am still triaging @Mathnerd314's commits: the guaranteed good ones I should probably rebase on top of my guaranteed good ones.

**Comment by Ericson2314** [comment]
Some decisions like "should `impure.nix` be part of `./default.nix`" don't affect complexity one way or the other and our purely stylistic. Whoever has authority to merge this should comment on those (I am guessing @edolstra, as this lies at the heart of nixpkgs).

File: default.nix [hunk 7]
```diff
@@ -7,33 +7,31 @@ if ! builtins ? nixVersion || builtins.compareVersions requiredVersion builtins.
...
-    let config_ = config; in # rename the function arguments
+  args @ { system ? builtins.currentSystem
+         , config ?
+            # The contents of the configuration file found at $NIXPKGS_CONFIG or
+            # $HOME/.nixpkgs/config.nix.
+            # for NIXOS (nixos-rebuild): use nixpkgs.config option to set
+            let
+              toPath = builtins.toPath;
+              getEnv = x: if builtins ? getEnv then builtins.getEnv x else "";
+              pathExists = name:
```
Comment (zimbatm) [inline default.nix]: toPath and pathExists are available since nix 0.10

Comment (Ericson2314) [inline default.nix]: Fixed, thanks!

File: pkgs/top-level/default.nix (hunk 6 above)
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: Fixed, thanks!

File: default.nix (hunk 7 above)
Comment (zimbatm) [inline default.nix]: Also see bec28d748c854abda373ca38831f4e77bc276fc1 in master

**Comment by Ericson2314** [comment]
Breaking this up to get things going with https://github.com/NixOS/nixpkgs/pull/16384 and https://github.com/NixOS/nixpkgs/pull/16382 being the first batch of trivial PRs.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/pull/16416 and https://github.com/NixOS/nixpkgs/pull/16415

**Comment by Ericson2314** [comment]
Ok, I finished triaging @mathnerd's commits. Everything is either already merged, squashed into my commits, or left to be dealt with after (see my `cross-rest` branch). I'm now trying to get travis to pass this---I fixed some legitimate bugs, and the current the current failure is due to somebody else's recent commit.

In light of all this, I am stripping the `semi-wip` prefix.

**Comment by Ericson2314** [comment]
I don't know how many people we need to review a change like this, but @vcunat, you mentioned you were interested at one point. Now would be a good time to take a look.

**Comment by Ericson2314** [comment]
Passes!

File: pkgs/top-level/default.nix [hunk 8]
```diff
@@ -1,36 +1,43 @@
-/* This file composes the Nix Packages collection.  That is, it
-   imports the functions that build the various packages, and calls
-   them with appropriate arguments.  The result is a set of all the
-   packages in the Nix Packages collection for some particular
-   platform. */
+/* This file composes a the Nix Packages collection. It:
```
Comment (joachifm) [inline pkgs/top-level/default.nix]: "a the" looks like a typo

Comment (Ericson2314) [inline pkgs/top-level/default.nix]: Indeed it is. Thanks!

**Comment by vcunat** [comment]
I've been severely out of time lately, and it's likely to last for weeks still. I'm sorry.

**Comment by Ericson2314** [comment]
That's ok. Thanks for saying so.

**Comment by Ericson2314** [comment]
@viric, @edolstra thought you might be a good person to review this?

[The merge conflict is trivial, but I am waiting for channels before rebasing so I can get a meaningful travis result.]

**Comment by copumpkin** [comment]
I like where this is going 👍

You might also be interested in the related (though possibly more ambitious) https://github.com/NixOS/nixpkgs/issues/11486

**Comment by Ericson2314** [comment]
Normal packages do have their hashes preserved, but I'm getting missing `isFreeBSD` flags for some reason on `xburstutils`, one of the few (the only?) users of cross toolchains in normal uncross nixpkgs. 

@copumpkin the last commit is my attempt to make debugging stdenvs with this new system easier, but I'm worried it just over-complicates. As someone who's definitely put blood sweat and tears into debugging that :), what do you think?

**Comment by Ericson2314** [comment]
Ok fixed (i.e. preserved the weirdness of) cross compiling--this PR should not change the semantics of that.

**Comment by Ericson2314** [comment]
@nbp Ok ready for review! After breaking out all those PRs, it will be nice to merge the last bits of the original!

**Comment by Ericson2314** [comment]
@nbp ping

File: pkgs/top-level/default.nix [hunk 9]
```diff
@@ -58,8 +69,14 @@ in let
...
-    inherit lib nixpkgsFun config platform;
+  allPackages = args: import ./stage.nix ({
+    inherit lib nixpkgsFun config;
   } // args);
 
+  inherit (import ../stdenv {
+    inherit lib allPackages system platform crossSystem config;
+  }) stdenv;
+
+  pkgs = allPackages { inherit system stdenv config crossSystem platform; };
```
Comment (nbp) [inline pkgs/top-level/default.nix]: Note:
This commit changes the dependencies of `stdenv`, and clean-up the stdenv story by removing the `defaultStdenv` attribute as well as the `bootStdenv` argument.

Before, stdenv was provided by all-packages, which was iterating multiple time over the `top-level/default.nix` expression.  This commit changes this relation such that the `stdenv` used for building all the *final* packages depends on the `allPackages` function, which does a fix-point over the set of packages needed for the bootstrap.  And the *final* set of packages depends on the `allPackages` function and on the `stdenv` result.

File: pkgs/stdenv/freebsd/default.nix [hunk 10]
```diff
@@ -5,7 +5,7 @@
 }:
 
 rec {
-  allPackages = import ../../..;
+  inherit allPackages;
```
Comment (nbp) [inline pkgs/stdenv/freebsd/default.nix]: nit: this modification should be part of the previous commit.

File: pkgs/stdenv/cross/default.nix [hunk 11]
```diff
@@ -1,14 +1,14 @@
...
 rec {
-  argClobber = {
+  vanillaStdenv = (import ../. {
+    inherit lib allPackages system platform;
     crossSystem = null;
     # Ignore custom stdenvs when cross compiling for compatability
     config = builtins.removeAttrs config [ "replaceStdenv" ];
-  };
-  vanillaStdenv = (import ../. (args // argClobber // {
-    allPackages = args: allPackages (argClobber // args);
```
Comment (nbp) [inline pkgs/stdenv/cross/default.nix]: Note:
This is fine, because, as the stdenv are now explicitly forced to provide a `config` attribute, by the fact that this commit removes the `config` value from `top-level/default.nix`. This `config` attribute would necessary be the `config` attribute given as argument to the `vanillaStdenv`.  So there is no need to make a custom version of `allPackages` here.

File: pkgs/top-level/platforms.nix [hunk 12]
```diff
@@ -443,4 +443,12 @@ rec {
     };
   };
 
+  chooseDefault = { system }:
+    if system == "armv6l-linux" then raspberrypi
```
Comment (nbp) [inline pkgs/top-level/platforms.nix]: nit: rename `chooseDefault` by `selectPlatformBySystem`.  No need to wrap the argument in an attribute set.

File: pkgs/stdenv/debug.nix [hunk 13]
```diff
@@ -0,0 +1,27 @@
+# This file mimicks top-levevl/default.nix, and is used to provide default
```
Comment (nbp) [inline pkgs/stdenv/debug.nix]: typo: `top-levevl`  --> `top-level`.

File: pkgs/stdenv/native/default.nix [hunk 14]
```diff
@@ -1,4 +1,9 @@
-{ system, allPackages ? import ../../.., config }:
+{ _defaults    ? import ../debug.nix
+, argsResolved ? _defaults.extend (_: _: args)
+, allPackages  ? argsResolved.allPackages
```
Comment (nbp) [inline pkgs/stdenv/native/default.nix]: I honestly think the code is better without these debug code.
This debug code makes it much more harder to understand what is going on with the argument.

I think if you are an stdenv developer adding a few `builtins.trace` function call is probably the easiest.
As well as making a wrapper file to invoke the right stage of the bootstrap you are looking for.

**Review by nbp — CHANGES_REQUESTED** [review]
Remove the last commit, and fix the nits, and I think this would be ok for merging.

File: pkgs/top-level/default.nix [hunk 15]
```diff
@@ -72,75 +57,9 @@ let
   # deterministically inferred the same way.
   nixpkgsFun = newArgs: import ./. (args // newArgs);
```
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: There should be comments to explain this, but actually I think this is OK. `config` is fine to `include`, but platform actually shouldn't be included because this function is usually called with a non-native `system`, which causes a different `platform` to be inferred. It would be bad to force the native `system`'s inferred `platform`.

Comment (Ericson2314) [inline pkgs/top-level/default.nix]: There should be comments to explain whatever route is chosen, in any event.

File: pkgs/top-level/default.nix [hunk 16]
```diff
@@ -72,75 +57,9 @@ let
...
-  #
-  # Example:
-  #  pkgs.overridePackages (self: super: {
-  #    foo = super.foo.override { ... };
-  #  }
-  #
-  # The result is `pkgs' where all the derivations depending on `foo'
-  # will use the new version.
+  # Partially apply some args for building phase pkgs sets
+  pkgs = import ./stage.nix ({
```
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: Hmm? `stage.nix` makes only a *single* stage--I am using stage in the sense of bootstrapping stage, not "fix-extend" stage, I assume that is the  source of the confusion? I would like better terms to distinguish the two.

File: pkgs/top-level/stage.nix [hunk 17]
```diff
@@ -80,10 +80,9 @@ let
       ((config.packageOverrides or (super: {})) super);
 
   # The complete chain of package set builders, applied from top to bottom
-  toFix = lib.foldl' (lib.flip lib.extends) (self: {}) [
+  toFix = lib.foldl' (lib.flip lib.extends) (self: stdenvDefault) [
```
Comment (Ericson2314) [inline pkgs/top-level/stage.nix]: Sure

Comment (Ericson2314) [inline pkgs/top-level/stage.nix]: Doing that

File: pkgs/stdenv/freebsd/default.nix (hunk 10 above)
Comment (Ericson2314) [inline pkgs/stdenv/freebsd/default.nix]: Right. Thanks, I meant to do that and forget.

File: pkgs/stdenv/cross/default.nix (hunk 11 above)
Comment (Ericson2314) [inline pkgs/stdenv/cross/default.nix]: Precisely.

File: pkgs/top-level/default.nix (hunk 9 above)
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: Incorporated into the commit message. Thanks.

File: pkgs/stdenv/native/default.nix (hunk 14 above)
Comment (Ericson2314) [inline pkgs/stdenv/native/default.nix]: I agree in principle, but here are two headaches:

1. One of the nixpkgs tests is building darwin bootstrap tools and booting a stdenv with it, so I need a decent portion of this complexity just to make an `allPackages` argument suitable for that test to pass.

2. I beleive the goal wrt debugging is less tracing than being able to load a small subset up, potentially with different arguments. That's also why these files are big `rec { .. }` rather than `let .. in ..`.

@copumpkin I think you are the person to ask via both of these, and more broadly what to do with this final commit.

File: pkgs/top-level/default.nix (hunk 15 above)
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: Ok explained in `HEAD^`.

File: pkgs/top-level/default.nix (hunk 16 above)
Comment (Ericson2314) [inline pkgs/top-level/default.nix]: Ok, I did some `s/phase/stage/` which should help. We already use "phase' for `mkDerivation`, and "stage" for this.

**Comment by Ericson2314** [comment]
@nbp Do you have time to re-review? Basically I went the opposite course of stripping the all the debugging hand-holds. Differences are roughly:
 - No more default arguments for debugging
 - `crossSystem` no longer defaults to `null` per stage and must be explicitly provided (but this is analogous to `system` and `platform`)
 - The first commit now also makes `pkgs/stdenv/default.nix` just the default rather than an attrset of them all (the others were already not exposed by that commit)
 - `top-level/default.nix` can take a custom stdenv now---one could already override it with config but same goes for platform and that's manually specifiable too.

If you want to diff, github.com/Ericson2314/nixpkgs/tree/cross-debug is the old version, and `cross-*` are some other ideas I tossed around. (github's comparing of diverged branches is unfortunately totally broken, so you'd have to use the CLI.0

**Comment by Ericson2314** [comment]
OK, so its been many months. @nbp told me he doesn't have time to review again for a while, but this is exactly in the spirit of his last review. 

There was only one part of his review I had reservations about, pertaining to my last commit. While removing the debugging aids (default arguments never used normally) does reduce complexity, it also may make debugging harder, and stdenvs are already pain enough. To alleviate that, I rebased my old final commit on top of this and opened  https://github.com/NixOS/nixpkgs/pull/20889 with it. I really want to see this merged, so my hope is that other PR can be merged by anyone who is adversely impacted by the removal of debugging aids in this.

So, my plan is to wait a day or two and then merge this myself. **Speak now, or ~~forever hold your peace~~ merge https://github.com/NixOS/nixpkgs/pull/20889**.
